File: frontend/src/api/design.ts
import api from './client'
import type { RequirementAnalyzeResponse } from './requirement'

export interface ModuleDesign {
  name: string
  responsibility: string
  inputs: string[]
  outputs: string[]
}

export interface ApiDesign {
  name: string
  method: string
  path: string
  description: string
}

export interface DataEntity {
  name: string
  fields: string[]
}

export interface DesignGenerateResponse {
  architecture_style: string
  architecture_rationale: string
  modules: ModuleDesign[]
  apis: ApiDesign[]
  data_entities: DataEntity[]
  mermaid: string
  raw_model_output: string
  reasoning_content?: string | null
  mode: 'fast' | 'deep'
}

export interface DesignStreamEvent {
  type: 'status' | 'reasoning' | 'content' | 'result' | 'error' | 'done'
  content: string
  mode?: 'fast' | 'deep'
  result?: DesignGenerateResponse | null
}

export async function generateDesign(requirement: RequirementAnalyzeResponse, mode: 'fast' | 'deep') {
  const { data } = await api.post<DesignGenerateResponse>('/design/generate', { requirement, mode })
  return data
}

export async function streamGenerateDesign(
  requirement: RequirementAnalyzeResponse,
  mode: 'fast' | 'deep',
  onEvent: (event: DesignStreamEvent) => void,
) {
  const response = await fetch(`${import.meta.env.VITE_API_BASE_URL}/design/generate/stream`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ requirement, mode }),
  })

  if (!response.ok || !response.body) {
    throw new Error('流式设计生成失败')
  }

  const reader = response.body.getReader()
  const decoder = new TextDecoder('utf-8')
  let buffer = ''

  while (true) {
    const { done, value } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })
    const parts = buffer.split('\n\n')
    buffer = parts.pop() || ''

    for (const part of parts) {
      const line = part.trim()
      if (!line.startsWith('data: ')) continue
      const payload = line.slice(6)
      onEvent(JSON.parse(payload) as DesignStreamEvent)
    }
  }
}

File: backend/app/schemas/design.py
from typing import Literal

from pydantic import BaseModel

from app.schemas.requirement import RequirementAnalyzeResponse


class DesignGenerateRequest(BaseModel):
    requirement: RequirementAnalyzeResponse
    mode: Literal["fast", "deep"] = "fast"


class ModuleDesign(BaseModel):
    name: str
    responsibility: str
    inputs: list[str]
    outputs: list[str]


class ApiDesign(BaseModel):
    name: str
    method: str
    path: str
    description: str


class DataEntity(BaseModel):
    name: str
    fields: list[str]


class DesignGenerateResponse(BaseModel):
    architecture_style: str
    architecture_rationale: str
    modules: list[ModuleDesign]
    apis: list[ApiDesign]
    data_entities: list[DataEntity]
    mermaid: str
    raw_model_output: str
    reasoning_content: str | None = None
    mode: Literal["fast", "deep"] = "deep"


class DesignStreamEvent(BaseModel):
    type: Literal["status", "reasoning", "content", "result", "error", "done"]
    content: str = ""
    mode: Literal["fast", "deep"] | None = None
    result: DesignGenerateResponse | None = None
